mqtt/mqtt-bridge/src/persist/publication_store.rs

```diff
@@ -25,44 +30,68 @@ impl PublicationStore<WakingMemoryStore> {
 
 impl<S: StreamWakeableState> PublicationStore<S> {
     pub fn new(state: S, batch_size: usize) -> Self {
-        let state = Arc::new(Mutex::new(state));
-        let loader = MessageLoader::new(Arc::clone(&state), batch_size);
-        let loader = Arc::new(Mutex::new(loader));
+        let state = Rc::new(RefCell::new(state));
+        let loader = MessageLoader::new(state.clone(), batch_size);
+        let loader = Rc::new(RefCell::new(loader));
 
         let offset = 0;
-        Self {
+        let inner = PublicationStoreInner {
             state,
             offset,
             loader,
-        }
+        };
+        let inner = Rc::new(RefCell::new(inner));
+
+        Self { inner }
     }
 
-    pub fn push(&mut self, message: Publication) -> Result<Key, PersistError> {
+    pub fn push(&self, message: Publication) -> Result<Key, PersistError> {
+        let mut inner = self.inner.borrow_mut();
+
         debug!(
             "persisting publication on topic {} with offset {}",
-            message.topic_name, self.offset
+            message.topic_name, inner.offset
         );
 
         let key = Key {
-            offset: self.offset,
+            offset: inner.offset,
         };
 
-        let mut state_lock = self.state.lock();
-        state_lock.insert(key, message)?;
-        self.offset += 1;
+        let mut state_borrow = inner
+            .state
+            .try_borrow_mut()
+            .map_err(PersistError::BorrowSharedState)?;
+        state_borrow.insert(key, message)?;
+        drop(state_borrow);
+
+        inner.offset += 1;
         Ok(key)
     }
 
-    pub fn remove(&mut self, key: Key) -> Result<(), PersistError> {
+    pub fn remove(&self, key: Key) -> Result<(), PersistError> {
+        let inner = self.inner.borrow_mut();
+
         debug!("removing publication with key {:?}", key);
 
-        let mut state_lock = self.state.lock();
-        state_lock.remove(key)?;
+        let mut state_borrow = inner
+            .state
+            .try_borrow_mut()
+            .map_err(PersistError::BorrowSharedState)?;
```

Comment:
We might want to panic and fail loud here. If there is an error then it is a significant logic error and we want to be able to easily see the effect of it.

Reply:
Considering that in the pump, the best way to handle borrow is to not panic, but shut down the pump cleanly, I would feel weird being inconsistent and panicking only in this case.